### PR TITLE
[cpu] add VecConvert between 8bits and 16bits

### PR DESCRIPTION
**Pull request by Valentine233**
The perf benefit was found in https://github.com/pytorch/pytorch/issues/124697#issuecomment-2071658300.

The PR adds intrinsic specializations between int8/uint8 and bf16/fp16.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/cpu/vec/vec256/vec256_convert.h [hunk 1]
```diff
@@ -128,30 +128,102 @@ struct VecConvert<int32_t, 1, uint8_t, 1> {
...
+        void>> {
+  static inline VectorizedN<Half, 1> apply(const VectorizedN<src_t, 1>& src) {
+    VectorizedN<float, 1> value = convert_int8_to_float<src_t>(src[0]);
+    VectorizedN<Half, 1> result;
+    result[0] = _mm256_castsi128_si256(cvtfp32_fp16(value[0]));
+    return result;
+  }
+};
+
+template <typename dst_t>
```
Comment (Valentine233) [inline aten/src/ATen/cpu/vec/vec256/vec256_convert.h:201]: Just some clang-format modifications for conversions between 8bits and float.

File: aten/src/ATen/cpu/vec/vec512/vec512_convert.h [hunk 2]
```diff
@@ -119,30 +119,102 @@ struct VecConvert<int32_t, 1, uint8_t, 1> {
...
+        void>> {
+  static inline VectorizedN<Half, 1> apply(const VectorizedN<src_t, 1>& src) {
+    VectorizedN<float, 1> value = convert_int8_to_float<src_t>(src[0]);
+    VectorizedN<Half, 1> result;
+    result[0] = _mm512_castsi256_si512(cvtfp32_fp16(value[0]));
+    return result;
+  }
+};
+
+template <typename dst_t>
```
Comment (Valentine233) [inline aten/src/ATen/cpu/vec/vec512/vec512_convert.h:192]: Just some clang-format modifications for conversions between 8bits and float.

File: aten/src/ATen/cpu/vec/vec256/vec256_convert.h [hunk 3]
```diff
@@ -128,30 +128,102 @@ struct VecConvert<int32_t, 1, uint8_t, 1> {
...
+        std::is_same_v<dst_t, unsigned char> ||
+            std::is_same_v<dst_t, signed char>,
+        void>> {
   static inline VectorizedN<dst_t, 1> apply(
-      const VectorizedN<float, 1>& src) {
-    return convert_float_to_int8<dst_t>(src[0]);
+      const VectorizedN<BFloat16, 1>& src) {
+    __m256 value;
+    cvtbf16_fp32(_mm256_castsi256_si128(src[0]), value);
+    return convert_float_to_int8<dst_t>(value);
```
Comment (jgong5) [inline aten/src/ATen/cpu/vec/vec256/vec256_convert.h:143]: why not calling the existing convert from int8 to fp32 and fp32 to bf16/fp16? Then you can combine bf16 and fp16 into a single implementation.

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: aten/src/ATen/cpu/vec/vec256/vec256_convert.h (hunk 3 above)
Comment (Valentine233) [reply aten/src/ATen/cpu/vec/vec256/vec256_convert.h]: Thanks for your advice. Modified it!

**Review by jgong5 — CHANGES_REQUESTED** [review]
Why do you change vec_base.h?

**Comment by Valentine233** [comment]
> Why do you change vec_base.h?

Just added a function `is_bit8` in `vec_base.h` to simplify codes. Sorry for not realizing that `clang-format` brought such a big code change. I have removed the irrelevant change now, thanks.

File: aten/src/ATen/cpu/vec/vec_base.h [hunk 4]
```diff
@@ -90,6 +90,16 @@ struct is_reduced_floating_point:
 template <typename T>
 constexpr bool is_reduced_floating_point_v = is_reduced_floating_point<T>::value;
 
+template <typename T>
+struct is_bit8:
```
Comment (jgong5) [inline aten/src/ATen/cpu/vec/vec_base.h:94]:
```suggestion
struct is_8bit_integer:
```

File: aten/src/ATen/cpu/vec/vec256/vec256_convert.h [hunk 5]
```diff
@@ -126,32 +126,48 @@ struct VecConvert<int32_t, 1, uint8_t, 1> {
...
-    void>> {
-  static inline VectorizedN<dst_t, 1> apply(
-      const VectorizedN<float, 1>& src) {
+    dst_t,
+    1,
+    float,
+    1,
+    typename std::enable_if_t<
+        std::is_same_v<dst_t, unsigned char> ||
+            std::is_same_v<dst_t, signed char>,
```
Comment (jgong5) [inline aten/src/ATen/cpu/vec/vec256/vec256_convert.h:153]: do you want to simplify similar checks like this too?

Comment (Valentine233) [reply aten/src/ATen/cpu/vec/vec256/vec256_convert.h]: Sure!

**Review by jgong5 — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**